src/main/java/com/neilbaner/duke/Duke.java

```diff
@@ -0,0 +1,174 @@
+        return Integer.parseInt(input.substring(5)) - 1;
+    }
+
+    private static int getIndexToDelete(String input) {
+        return Integer.parseInt(input.substring(7)) - 1;
+    }
+
+    private static String getDeadlineTitle(String input) {
+        return input.substring(9, input.indexOf("/by")).strip();
+    }
+
+    private static void saveState() throws IOException {
+        FileWriter fw = new FileWriter("dukesave.txt", false);
+        fw.write(Serializer.serializeTaskList(list));
+        fw.close();
+    }
+
+    private static boolean parseInput(String input) throws DukeException {
+        String lowerCaseInput = input.toLowerCase();
+        if (lowerCaseInput.equals("bye")) {
+            Messages.printGoodBye();
+            try{
+                saveState();
+            }catch (IOException e) {
+                System.err.println("Error saving the state. Your data was not saved.");
+            }
+            return false;
+        } else if (lowerCaseInput.equals("list")) {
+            printTaskList();
+        } else if (lowerCaseInput.startsWith("done")) {
+            try {
+                int indexToMark = getIndexToMark(input);
+                markAsDone(indexToMark);
+            } catch (NumberFormatException e) {
+                Messages.printFormattingError();
+            } catch (IndexOutOfBoundsException e) {
+                throw new TaskIndexOutOfBoundsException();
+            }
+        } else if (lowerCaseInput.startsWith("delete")) {
+            try {
+                int indexToDelete = getIndexToDelete(input);
+                deleteTask(indexToDelete);
+            } catch (NumberFormatException e) {
+                Messages.printFormattingError();
+            } catch (IndexOutOfBoundsException e) {
+                throw new TaskIndexOutOfBoundsException();
+            }
+        } else if (lowerCaseInput.startsWith("todo")) {
+            addToDo(input.substring(5));
+        } else if (lowerCaseInput.startsWith("deadline")) {
+            try {
+                String title = getDeadlineTitle(input);
+                String dueDate = getDeadlineDueDate(input);
+                if (dueDate.equals("")) {
+                    throw new BlankDeadlineDateException();
+                }
+                addDeadline(title, dueDate);
+            } catch (IndexOutOfBoundsException e) {
+                Messages.printFormattingError();
+            }
+        } else if (input.startsWith("event")) {
+            try {
+                String title = getEventTitle(input);
+                String eventTime = getEventTime(input);
+                if (eventTime.equals("")) {
+                    throw new BlankEventTimeException();
+                }
+                addEvent(title, eventTime);
+            } catch (IndexOutOfBoundsException e) {
+                Messages.printFormattingError();
+            }
+        } else {
+            throw new UnknownCommandException();
```

Comment:
you might want to look at the situation when the user enters "    valid command"

Reply:
I'm sorry, but I'm not entirely sure what you mean; the output I get is "I'm sorry, I'm not sure what valid command means" followed by the help text (as expected). Do you mean that the sentence itself doesn't make sense and I should enclose the user input in double quotes to make it more clear?

Reply:
hi Neil, I mean something like "   list". The spaces infront of the command might not be recognized by your system.